Make native linux executor to accept extra environment variables

Fadis opened the pull request:
<!-- Set a title that summarizes the PR changes. -->

## Type of PR
<!-- This project uses issue-driven development, so please take the appropriate action for each PR type. -->

- [ ] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [x] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue
<!-- If this PR is a PR type A/B-1, please provide a link to the corresponding issue. -->
<!-- If this PR is a PR type B-1, please write "N/A" -->

## Importance of PR
<!-- Please describe the importance of the PR in terms of the following aspects. -->

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [ ] Medium (requires several hours to a day of review)
    - [x] Small (trivial changes, such as build error)

## PR Overview

This change append new constructor argument to NativeLinuxExecutor. This make possible to set environment variables on specific child process.
Since the argument is defaulted to {}, There is no need to migrate existing use of NativeLinuxExecutor that doesn't use special environment variables.

<!-- Please provide a summary of this PR. -->
<!-- If this PR is a PR type A/B-1, this PR will be considered as an item in the checklist for the related issue. Please provide a link to the issue comment that contains the checklist. -->
<!-- If this PR is a PR type B-2, unnecessary to reference the issue. Please provide a summary. -->

## Concerns (Optional)
<!-- If you have any concerns, please describe them clearly by filling in the relevant checklist items below. If there is anything else you would like to share with the reviewer, please include it. -->

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [ ] The PR title is a summary of the changes.
- [ ] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

Commit:
Make native linux executor to accept extra environment variables

potetisensei commented on executor/native_linux_executor.cpp:
```diff
@@ -485,9 +487,14 @@ void NativeLinuxExecutor::Run(const u8 *buf, u32 len, u32 timeout_ms) {
                 dup2(null_fd, 0);
             }
+	    std::vector< char* > temporal_envp;
             // Execute new executable binary on the child process.
 	    // If failed, that matter is recorded to child_state.
-            child_state->exec_result = execv(cargv[0], (char**)cargv.data());
+            child_state->exec_result = execve(
+                cargv[0],
+                (char**)cargv.data(),
+                GetJoinedEnvironmentVariables( temporal_envp )
+	    );
```
While calling `GetJoinedEnvironmentVariables` here is totally valid, I think it is a bit naive to create a `std::vector<char*>` instance every time.

How about makng `NativeLinuxExecutor` have `std::vector<char*> temporal_envp` as a private member variable like `NativeLinuxExecutor::cenvp`? It's just the same as `Executor::cargv`. Because the lifetime of `cenvp` would be the same as that of `environment_variables`, I don't see it so problematic.

Fadis replied:
fixed

potetisensei commented on executor/native_linux_executor.cpp:
```diff
@@ -67,7 +67,8 @@ NativeLinuxExecutor::NativeLinuxExecutor(
     u32  bb_shm_size,
-    bool record_stdout_and_err
+    bool record_stdout_and_err,
+    fuzzuf::executor::EnvironmentVariables &&environment_variables_
```
In my opinion, it is easier to imagine what `environment_variables_` actually is if we directly use `std::vector<std::string>` instead of the type alias `fuzzuf::executor::EnvironmentVariables`. 
This is because, when we see the unseen type `EnvironmentVariables`, we have to investigate what it is by searching its definition. 

Are you going to change the definition of `EnvironmentVariables` somehow later? 
Like transforming it to `std::map<std::string, std::string>` or defining a new struct?
If not, I don't think the type alias is needed.

Fadis replied:
fixed

potetisensei reviewed (changes requested):
This PR would be needed to add some new fuzzing algorithms. 
The diffs of the PR are also almost clear and appropriate.
I left only two requests, so please check them.

> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [x] The reviewer assigned more reviewers if needed.
- [x] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [x] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.

Commit:
Avoid creating joined environment variables for each execution of NativeLinuxExecutor

CAUTION : This is a destructive change. setenv will no longer affect to executor created prioir.
* The change make NativeLinuxExecutor to cache joined environment variables.

potetisensei commented on include/fuzzuf/executor/native_linux_executor.hpp:
```diff
@@ -149,5 +143,6 @@ class NativeLinuxExecutor : public Executor {
     bool record_stdout_and_err;
-    fuzzuf::executor::EnvironmentVariables environment_variables;
+    std::vector< std::string > environment_variables;
+    std::vector< char* > raw_environment_variables;
```
Would you leave a comment like "raw_environment_variables := envrion + [s.c_str() for s in environment_variables]" here?

Fadis replied:
fixed

potetisensei reviewed (approved):
LGTM

> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [x] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [x] If minor refactorings are not mentioned in the PR, I understand the intent.
- [x] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [x] The reviewer understands how this PR addresses the issue and the specific changes.
- [x] This PR uses the best possible issue resolution that the reviewer can think of.
- [x] This PR is ready to be merged.
